File: frontend/lib/slugs.ts
// frontend/lib/slugs.ts
/**
 * SEO-friendly URL slug utilities (Shopify-style)
 * Generates and parses product slugs like: "/produkte/rinos-gaia-2-gravel-bike-rinos24grx400"
 */

/**
 * Generate SEO-friendly slug from product name and article number
 * Example: ("RINOS Gaia 2 Gravel Bike", "RINOS24GRX400") -> "rinos-gaia-2-gravel-bike-rinos24grx400"
 */
export function generateProductSlug(productName: string, articlenr: string): string {
  if (!productName) {
    return articlenr.toLowerCase()
  }

  // Normalize unicode characters (ö -> o, ä -> a, ü -> u, ß -> ss)
  let slug = productName.normalize('NFKD').replace(/[\u0300-\u036f]/g, '')

  // Convert to lowercase
  slug = slug.toLowerCase()

  // Remove any characters that aren't alphanumeric, spaces, or hyphens
  slug = slug.replace(/[^\w\s-]/g, '')

  // Replace spaces and underscores with hyphens
  slug = slug.replace(/[\s_]+/g, '-')

  // Remove multiple consecutive hyphens
  slug = slug.replace(/-+/g, '-')

  // Strip hyphens from start and end
  slug = slug.trim().replace(/^-+|-+$/g, '')

  // Append article number for uniqueness (Shopify-style)
  if (articlenr) {
    slug = `${slug}-${articlenr.toLowerCase()}`
  }

  return slug
}

/**
 * Extract article number from a product slug
 * Example: "rinos-gaia-2-gravel-bike-rinos24grx400" -> "RINOS24GRX400"
 * Falls back to the input if it looks like an article number already
 */
export function extractArticleNrFromSlug(slug: string): string {
  // If it's already an article number (uppercase with numbers), return as-is
  if (/^[A-Z0-9]+$/i.test(slug)) {
    return slug.toUpperCase()
  }

  // Extract the last segment (should be the article number)
  const parts = slug.split('-')
  const lastPart = parts[parts.length - 1]

  // Look for article number pattern (letters + numbers)
  // Try to find the article number in the last few segments
  for (let i = parts.length - 1; i >= Math.max(0, parts.length - 3); i--) {
    const segment = parts.slice(i).join('')
    // Check if this looks like an article number (mix of letters and numbers)
    if (/^[a-z]+\d+[a-z]*\d*$/i.test(segment)) {
      return segment.toUpperCase()
    }
  }

  // Fallback: return the slug as-is (uppercase)
  return slug.toUpperCase()
}

/**
 * Generate category slug
 * Example: "Gravel Bikes" -> "gravel-bikes"
 */
export function generateCategorySlug(categoryName: string): string {
  return generateProductSlug(categoryName, '')
}

File: backend/utils/slugs.py
# backend/utils/slugs.py
"""
URL slug generation utilities for SEO-friendly URLs (Shopify-style)
"""
import re
import unicodedata


def generate_slug(text: str, articlenr: str = None) -> str:
    """
    Generate SEO-friendly URL slug from product name

    Examples:
        - "RINOS Gaia 2 Gravel Bike" + "RINOS24GRX400" -> "rinos-gaia-2-gravel-bike-rinos24grx400"
        - "RINOS Sandman 6.0 Gravelbike" + "SANDMAN6" -> "rinos-sandman-60-gravelbike-sandman6"

    Args:
        text: Product name or text to slugify
        articlenr: Optional article number to append for uniqueness

    Returns:
        URL-safe slug string
    """
    if not text:
        return articlenr.lower() if articlenr else ""

    # Normalize unicode characters (ö -> o, ä -> a, ü -> u, ß -> ss, etc.)
    text = unicodedata.normalize('NFKD', text)
    text = text.encode('ascii', 'ignore').decode('ascii')

    # Convert to lowercase
    text = text.lower()

    # Remove any characters that aren't alphanumeric, spaces, or hyphens
    text = re.sub(r'[^\w\s-]', '', text)

    # Replace spaces and underscores with hyphens
    text = re.sub(r'[\s_]+', '-', text)

    # Remove multiple consecutive hyphens
    text = re.sub(r'-+', '-', text)

    # Strip hyphens from start and end
    text = text.strip('-')

    # Append article number for uniqueness (Shopify-style)
    if articlenr:
        text = f"{text}-{articlenr.lower()}"

    return text


def generate_category_slug(category_name: str) -> str:
    """
    Generate SEO-friendly slug for category pages

    Examples:
        - "Gravel Bikes" -> "gravel-bikes"
        - "E-Bikes & Pedelecs" -> "e-bikes-pedelecs"

    Args:
        category_name: Category name to slugify

    Returns:
        URL-safe category slug
    """
    return generate_slug(category_name)
